pallets/poa/src/lib.rs

```diff
@@ -0,0 +1,710 @@
+        /// of validators don't exceed the max allowed count. The validator is considered for adding at
+        /// the end of this epoch unless `short_circuit` is true. If a validator is already added to the queue
+        /// an error will be thrown unless `short_circuit` is true, in which case it swallows the error.
+        // Weight can be 0 as its called by Master. TODO: Use signed extension to make it free
+        #[weight = 0]
+        pub fn add_validator(origin, validator_id: T::AccountId, short_circuit: bool) -> dispatch::DispatchResult {
+            // TODO: Check the origin is Master
+            ensure_root(origin)?;
+
+            Self::add_validator_(validator_id, short_circuit)
+        }
+
+        /// Remove the given validator from active validator set and the queued validators at the end
+        /// of epoch unless `short_circuit` is true. If validator is already queued for removal, an error
+        /// will be thrown unless `short_circuit` is true, in which case it swallows the error.
+        /// It will not remove the validator if the removal will cause the active validator set to
+        /// be empty even after considering the queued validators.
+        // Weight can be 0 as its called by Master. TODO: Use signed extension to make it free
+        #[weight = 0]
+        pub fn remove_validator(origin, validator_id: T::AccountId, short_circuit: bool) -> dispatch::DispatchResult {
+            // TODO: Check the origin is Master
+            ensure_root(origin)?;
+            Self::remove_validator_(validator_id, short_circuit)
+        }
+
+        /// Replace an active validator (`old_validator_id`) with a new validator (`new_validator_id`)
+        /// without waiting for epoch to end. Throws error if `old_validator_id` is not active or
+        /// `new_validator_id` is already active. Also useful when a validator wants to rotate his account.
+        // Weight can be 0 as its called by Master. TODO: Use signed extension to make it free
+        #[weight = 0]
+        pub fn swap_validator(origin, old_validator_id: T::AccountId, new_validator_id: T::AccountId) -> dispatch::DispatchResult {
+            // TODO: Check the origin is Master
+            ensure_root(origin)?;
+            Self::swap_validator_(old_validator_id, new_validator_id)
+        }
+
+        /// Awards the complete txn fees to the block author if any and increment block count for
+        /// current epoch and who authored it.
+        fn on_finalize() {
+            print("Finalized block");
+
+            // Get the current block author
+            let author = <pallet_authorship::Module<T>>::author();
+
+            Self::award_txn_fees_if_any(&author);
+
+            Self::increment_current_epoch_block_count(author)
+        }
+    }
+}
+
+impl<T: Trait> Module<T> {
+    fn add_validator_(validator_id: T::AccountId, short_circuit: bool) -> dispatch::DispatchResult {
```

Comment:
It's apparent that this function is two functions, `add_vailidator_(id, false)` and `add_validator_(id, true)`.

`add_vailidator_(id, false)` enqueues an account to be added to the validator set, failing if the account is already in the queue. The queue will be processed at the end of the session.

`add_vailidator_(id, true)` places an account at the front of the queue, regardless of whether it is in the queue already. After placing the account at the front of the queue, it orders a session change at the next fair time. At that session change, the account will be popped from the queue but only if there is room for it in the validator set. At the session change any actions previously ordered during this session (e.g. previous calls to add_validator_(id, false)) will also be executed.

Another way to view `short_circuit` is as the conflation of three options:
1. should validator be added to the front, or back of the queue,
2. after adding validator to queue, should short_circuit_current_epoch() be called, and
3. should it be allowed to re-add validator to the queue

Here is my initial suggestion for simplifying this function: remove the `short_circuit` flag and replace it with a `QueuePosition` enum. Add a new root-only function `end_session_fairly()` that calls `short_circuit_current_epoch()`.

Reply:
Passing `true` will not re-add, it either adds it to the front of the queue or moves it to the front. If adding `end_session_fairly`, what other functions are added or removed or their params changed

Reply:
```rust
fn add_validator(T::AccountId, QueuePosition) -> DispatchResult;
fn remove_validator(T::AccountId) -> DispatchResult;
fn swap_validator(T::AccountId, T::AccountId) -> DispatchResult;
fn end_session_fairly() -> DispatchResult;
fn end_session_asap() -> DispatchResult;
```

Reply:
Alternatively, if you want to keep things simple:

```rust
/// always calls end_session_fairly after runnning
fn modify_validators(additions: BTreeSet<T::AccountId>, removals: BTreeSet<T::AccountId>) -> DispatchResult;
fn end_session_asap() -> DispatchResult;
```

Reply:
This solution seems nicest to me at the moment:

```rust
fn modify_validators(
    additions: BTreeSet<T::AccountId>,
    removals: BTreeSet<T::AccountId>,
    when: EndSession,
) -> DispatchResult;

enum EndSession {
    OnSchedule,
    Fairly,
    Asap,
}
```

Reply:
Agree that it can be done this way. But this would mean changing lot of existing code and tests. So skipping for now